private/bufpkg/bufimage/bufimageutil/bufimageutil.go

```diff
@@ -266,27 +266,39 @@ func ImageFilteredByTypesWithOptions(image bufimage.Image, types []string, opts
 		// the file's WeakDependency field.
 		indexFromTo := make(map[int32]int32)
 		indexTo := 0
+		// Only handle imports and dependencies if there are any.
 		for indexFrom, importPath := range imageFileDescriptor.GetDependency() {
 			path := append(basePath, int32(indexFrom))
-			if _, ok := importsRequired[importPath]; ok {
-				sourcePathRemapper.markMoved(path, int32(indexTo))
-				indexFromTo[int32(indexFrom)] = int32(indexTo)
-				imageFileDescriptor.Dependency[indexTo] = importPath
-				indexTo++
-				// markDeleted them as we go, so we know which ones weren't in the list
-				delete(importsRequired, importPath)
-			} else {
+			// If there are no required imports, we mark this as deleted.
+			if importsRequired == nil {
 				sourcePathRemapper.markDeleted(path)
+			} else {
+				// Otherwise, we check if the import path exists among required imports. If yes, we
+				// move and then delete from required imports as we go.
+				if i := importsRequired.index(importPath); i != -1 {
+					sourcePathRemapper.markMoved(path, int32(indexTo))
+					indexFromTo[int32(indexFrom)] = int32(indexTo)
+					imageFileDescriptor.Dependency[indexTo] = importPath
+					indexTo++
+					// delete them as we go, so we know which ones weren't in the list
+					if err := importsRequired.delete(importPath); err != nil {
+						return nil, err
+					}
+				} else {
+					// Path did not exist in required imports, we mark as deleted.
+					sourcePathRemapper.markDeleted(path)
+				}
```

Comment:
Still think this could be moved back to:
```go
if importsRequired != nil && importsRequired.index(importPath) != -1 { ... } else { ... }
```
As the branch is the has check but for nil and existence. The value of `i` isn't used. So I think it makes sense to move into one.